src/Collector/ComponentCollector.php

```diff
@@ -73,6 +109,25 @@ public function processNode(Node $node, Scope $scope): ?array
         return null;
     }
 
+    /**
+     * @param Node[] $stmts
+     * @phpstan-return CollectedComponentArray[]
+     */
+    private function processStmts(array $stmts, Scope $scope): array
+    {
+        $return = [];
+        foreach ($stmts as $stmt) {
+            if (isset($stmt->stmts)) {
+                $return = array_merge($return, $this->processStmts($stmt->stmts, $scope));
+            }
+            $procesNodeReturn = $this->processNode($stmt, $scope);
```

Comment:
Are you sure Scope will have correct data for variable types etc? At first sight it seems you bypassed PHPStan logic that allows expression type resolving by Scope. You might have scope freezed at point of trait use used to analyse all content of trait.

Reply:
I believe it will work.
1) if I have scope for trait, I don’t have class reflection
2) traits are copied to classes in runtime
3) I think I saw it in phpstan itself to use classScope on traits

Reply:
Also I believe someone will figure out how to use phpstan built-in methods. I just don’t want to freeze at this point so long :)

Reply:
Problem is that you have frozen scope. youa re not updating it based on analyzed statements. In PHPStan sopce changes after every statement. For example after variable assign Scope remembers what type was assigned to variable so later when you ask scope to getTYpe of that variable it is known. This will not work in your setup I am afraid. Only variables known at the start will be known. Nothing that happens inside trait will update scope.

Reply:
This code in Trait will failed to resolve I am afraid:
```php
$var = 'string'; // <-- analysis of this statement will not update Scope
$this->emplate->myVariable = $var; // <-- VariableCollector would resolve myVariable as mixed because information that $var exists and is string will not be updated in Scope during trait analysis
```

Reply:
I see, and how to change scope? :) it's black magic for me. I was so excited it works… you calm down me :D

Reply:
Well that is what NodeScopeResolver do as far as I understand

Reply:
well I tried it and it can find types of variables, but it don't know in which method it is assigned...

Reply:
Hmm I wonder how information of variable type got into scope. What code you tested?

Reply:
I'm not sure now, maybe I hadn't this type of assign that time. You're right this is not working now.